### PR TITLE
CCTexture

### PR DESCRIPTION
**Pull request by slembcke**
This pull request replaces #1150 since the develop-v4 branch was deleted.

This pull requests the CCTexture specific changes proposed in this document:
https://docs.google.com/document/d/1p5uhsyshRRasGStdEi0_sdBuGgbUSS7BypOji75O2pE/edit?usp=sharing
It does not include serialization or render texture changes. Those will be separate tasks.

Texture filtering/addressing can now be specified without using (mostly broken) private APIs. The PVR file loader has been completely rewritten to support loading from CCFile objects (gzipped, encrypted, remote, etc). PVR support for Metal is also working, though currently only for RGBA8 format files. The Metal/GL code is better separated which cuts down on some of the #ifdef gobbledygook and should make improved threading support much easier later on. Lastly, cubemap textures were added for better Cocos3D interoperability, CCEffects, and custom shader use.

### CODE REVIEW

File: cocos2d/CCDirector.m [hunk 1]
```diff
@@ -970,16 +968,17 @@ -(void) createStatsLabel
 	NSData *nsdata = [NSData dataWithBytes:data length:data_len];
 	CGDataProviderRef imgDataProvider = CGDataProviderCreateWithCFData( (__bridge CFDataRef) nsdata);
 	CGImageRef imageRef = CGImageCreateWithPNGDataProvider(imgDataProvider, NULL, true, kCGRenderingIntentDefault);
-	CCTexture *texture = [[CCTexture alloc] initWithCGImage:imageRef contentScale:contentScale];
 	CGDataProviderRelease(imgDataProvider);
+    
+    CCImage *image = [[CCImage alloc] initWithCGImage:imageRef contentScale:contentScale options:nil];
```
Comment (andykorth) [inline cocos2d/CCDirector.m]: Aha, found a different indentation level! It's easier to find them in github than xcode.

File: cocos2d/CCLabelTTF.m [hunk 2]
```diff
@@ -645,289 +644,252 @@ - (CCTexture*) createTextureWithAttributedString:(NSAttributedString*)attributed
...
-
-#endif
+    CCImage *image = [[CCImage alloc] initWithPixelSize:POTSize contentScale:[CCDirector sharedDirector].contentScaleFactor pixelData:pixelData];
+    image.contentSize = CC_SIZE_SCALE(dimensions, 1.0/image.contentScale);
+	return [[CCTexture alloc] initWithImage:image options:nil];
+}
+
+
+//#pragma mark -
+//#pragma mark Render Font iOS 5
```
Comment (andykorth) [inline cocos2d/CCLabelTTF.m]: I vote for just removing this.

File: cocos2d/CCTMXXMLParser.m [hunk 3]
```diff
@@ -486,35 +482,23 @@ - (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName names
...
-		len = base64Decode((unsigned char*)[_currentString UTF8String], (unsigned int) [_currentString length], &buffer);
-		if( ! buffer ) {
-			CCLOG(@"cocos2d: TiledMap: decode data error");
-			return;
-		}
+        
+        NSData *decoded = [[NSData alloc] initWithBase64Encoding:_currentString];
+        NSAssert(decoded, @"Could not decode tilemap data.");
 
 		if( _layerAttribs & (TMXLayerAttribGzip | TMXLayerAttribZlib) ) {
```
Comment (andykorth) [inline cocos2d/CCTMXXMLParser.m:489]: Looks like more indentation oddities around here. Are they worth pointing out in code reviews?

Comment (slembcke) [inline cocos2d/CCTMXXMLParser.m:489]: Might as well.

File: cocos2d/CCLabelTTF.m (hunk 2 above)
Comment (slembcke) [inline cocos2d/CCLabelTTF.m]: That was meant to be removed actually. We stopped officially supporting iOS 5 sometime last summer?

**Comment by slembcke** [comment]
_ping_

This pull request is now 3 weeks old and important for finishing the new file utils integration.

**Comment by andykorth** [comment]
I looked over Scott's shoulder as it was being written, so it has my stamp of approval ;D

**Pull request merged**